src/java.base/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -286,6 +286,17 @@ public MemorySegment copyFrom(MemorySegment src) {
     @Override
     public long mismatch(MemorySegment other) {
         Objects.requireNonNull(other);
+        if (
+            other instanceof AbstractMemorySegmentImpl that &&
+            unsafeGetBase() == that.unsafeGetBase() &&
+            unsafeGetOffset() == that.unsafeGetOffset() &&
+            byteSize() == that.byteSize()
+        ) {
+            this.checkAccess(0, this.byteSize(), true);
+            that.checkAccess(0, that.byteSize(), true);
+            checkValidState();
+            return -1;
+        }
```

Comment:
I'm skeptical of trying to optimize this without a more thorough performance investigation. We shouldn't add this complexity on a whim.

I'm more in favor of just dropping the check from the static mismatch method and leaving it at that. (And adding a test as Maurizio mentioned).

Reply:
I have similar thoughts. This is low-level API. Code using it has more context knowledge and is more fit to do optimisations like that or not, depending if they are worth it. I would want such low level API to do as little checks as possible to accommodate faster common case scenario.